validate argument axis of tf.experimental.numpy.stack

SuryanarayanaY opened the pull request:
AT present `tf.experimental.numpy.stack` not validating the values for `axis` argument. If we pass out of bound values to the `axis` still it is not rising the Error. Hence I am proposing amendments to validate the `axis` argument and raise the Error in case of out of bound axis.

Please find the attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/db882df3c2245044a96cd943dead229e/git_55217_nightly-2-14.ipynb) for explaining the problem and also with proposed amendments to the code it works fine and raised intended error when invalid axis arguments provided.

Fixes #55217

SuryanarayanaY committed:
validate tf.experimental.numpy.stack axis argument

AT present tf.experimental.numpy.stack not validating the axis argument. If we pass out of bound values to the axis still it is not rising the Error. Hence I am proposing amendment to validate the axis argument and raise the Error in case of out of bound axis.

Please find the attached gist for explaining the problem and also with proposed amendments to the code it works fine and raised intended error when invalid axis arguments provided.

cantonios commented:
It _is_ raising an error.  Even in your gist you provide in this description.  So what's the issue?

SuryanarayanaY commented:
> It _is_ raising an error. Even in your gist you provide in this description. So what's the issue?

Sorry you might have gone through the `tf.stack`. I have deleted that part in gist and updated it.

Its not raising the error for `tensorflow.experimental.numpy.stack`. Please check the below code in gist for and its not raising the error. 

```
import tensorflow.experimental.numpy as tnp
tnp.experimental_enable_numpy_behavior()
x = tf.random.uniform([5])
tnp.stack(x, axis=-2) # Passes without any error
```

Thanks!

cantonios commented:
That's not throwing an error because the axis is valid: it can be in the range [0, 1], and negative values work backwards from the end, so -2 corresponds to axis=0 and -1 corresponds to axis=1 in this case.  Your example fails if you pass axis=-3.

gbaned commented:
Hi @SuryanarayanaY Can you please check @cantonios's comments ? Thank you!

SuryanarayanaY commented:
For earlier case the explanation holds right. I tested with new code.Please refer below code.

```
import tensorflow.experimental.numpy as tnp
tnp.experimental_enable_numpy_behavior()
x = tf.random.uniform([5,3])
tnp.stack(x, axis=-4) # passes without error
tnp.stack(x, axis=-3) # passes without error
```

Whereas with Numpy it raises error.

```
import numpy as np
x_n = np.random.randn(5, 3)
print(np.shape(x_n))
np.stack(x_n, axis=-4) # raises error
np.stack(x_n, axis=-3) # raises error
```
Please refer the new gist attached [here](https://colab.research.google.com/gist/SuryanarayanaY/294965c552daca8eb24f0189f865908f/60727.ipynb)  and the proposed code change might resolve this.

gbaned commented:
Hi @cantonios Can you please review this PR ? Thank you!

cantonios commented on tensorflow/python/ops/numpy_ops/np_array_ops.py:
```diff
@@ -1066,6 +1066,10 @@ def stack(arrays, axis=0):  # pylint: disable=missing-function-docstring
     arrays = asarray(arrays)
     if axis == 0:
       return arrays
+    elif axis < -array_ops.rank(arrays) or axis >= array_ops.rank(arrays):
```
The issue is actually in `swapaxes`, which doesn't properly check that the axis is in the valid range.

SuryanarayanaY replied:
@cantonios 

I have tested implementing range validation in swapaxes and it seems working fine. Attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/81f511dd5eb0ecb4adf82e3dfd1260e1/60727_pr-r1.ipynb) for reference.

Validation Code and location can be found below.

```
def swapaxes(a, axis1, axis2):  # pylint: disable=missing-docstring
  a = asarray(a) #
  def adjust_axes(axes, rank):
    def f(x):
      if isinstance(x, int):
        if x < 0:
          x = x + rank
      else:
        x = array_ops.where_v2(x < 0, np_utils.add(x, a_rank), x)
      return x
    return nest.map_structure(f, axes)

  if axis2 < -array_ops.rank(a) or axis2 >= array_ops.rank(a):
    raise ValueError(
        f"Argument `axis` = {axis2} not in range "
                      f"[{-array_ops.rank(a)}, {array_ops.rank(a)})")  
```

Is it good to go implementing the validation code in `swapaxes` instead of `stack` ?

cantonios replied:
The check should be in `swapaxes`, which has the same issue.  It currently assumes that `axis` is always in the range `[-rank, rank)`.

cantonios replied:
Still waiting for the check to be made in `swapaxes`

gbaned commented:
Hi @SuryanarayanaY  Can you please check @cantonios's comments and keep us posted ? Thank you!

SuryanarayanaY commented:
Requested for review

gbaned commented:
Hi @cantonios Can you please review this PR ? Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

SuryanarayanaY committed:
validated axis argument in swapaxes

validated axis argument in swapaxes for raising exception for invalid axis argument

SuryanarayanaY commented:
@gbaned , Could you please review and do needful. Thanks

cantonios commented on tensorflow/python/ops/numpy_ops/np_array_ops.py:
```diff
@@ -819,6 +819,11 @@ def f(x):
       return x
     return nest.map_structure(f, axes)
+  
+  if axis2 < -array_ops.rank(a) or axis2 >= array_ops.rank(a):
```
This fails if these are tensors, since you can't use a tensor as a python bool.   You probably need to use something like this: https://github.com/tensorflow/tensorflow/blob/517df2d259ba27f2ef3da399bf9c3b081d7a57b7/tensorflow/python/ops/image_ops_impl.py#L71

though we shouldn't add the dependency on image_ops_impl, so maybe re-implement it, or find another similar function in a shared utility

gbaned commented:
Hi @SuryanarayanaY Can you please check @cantonios's comments ? Thanks you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!
